Generate CMakeSettings.json which defines the Visual Studio build config

JoergAtGithub opened the pull request:
To build Mixxx in the Visual Studio IDE, it should be enough to do the following steps:
1.) Run:
tools\windows_buildenv.bat setup
2.) Open Visual Studio
3.) Open Local Folder and select the mixxx directory
4.) Menu Project -> Generate Cache
5.) Menu Build -> Build All

I added these build steps to README.md

JoergAtGithub committed:
Initial batch to generate CMakeSettings.json which defines the Visual Studio build environments

Be-ing commented:
Great idea. Thanks for working on this.

Holzhaus commented on tools/windows_buildenv.bat:
```diff
@@ -97,3 +102,38 @@ EXIT /B 0
+    >>%CMakeSettings% echo     }
+    >>%CMakeSettings% echo   ]
+    >>%CMakeSettings% echo }
+    GOTO :EOF
```
Missing newline at EOF. Please use our pre-commit hooks to fix these kind of issue automatically.

Holzhaus reviewed (approved):
Good idea, thanks.

poelzi commented:
I set -DCMAKE_EXPORT_COMPILE_COMMANDS=ON in the nix-shell environment which helps vscode to know how it builds a project and can display compile errors before you even start compiling. Maybe we should enable this in general ?

JoergAtGithub committed:
Enabled Clang-tidy and CMAKE_EXPORT_COMPILE_COMMANDS

JoergAtGithub committed:
Added CMake configurations for all Mixxx optimize levels
Used unique names for the build and dist directories of each configuration
Added generated files and directories to .gitignore
Used CMake variables per switch, instead of cmakeCommandArgs

JoergAtGithub committed:
windows_buildenv.bat can be now executed by double click
Old CMakeSettings.json is stored in backupfile instead of overwriting it

JoergAtGithub committed:
Updated the Compile section in README.md

JoergAtGithub commented:
Ready for review!

Be-ing commented on README.md:
```diff
@@ -31,21 +31,30 @@ Have a bug or feature request? [File a bug on Launchpad][fileabug].
+1. Download these sources (using git checkout as described in [Using Git])
+2. Run the batch file `tools\windows_buildenv.bat`
+   - This file downloads the prebuild Mixxx environment, defined in `cmake\windows_build_environment_name` from https://downloads.mixxx.org/builds/buildserver/2.3.x-windows/
+   - Generates the `CMakeSettings.json` with the matching build configurations for Visual Studio
```
Can you write instructions for building on the command line without coupling it to any particular IDE?

JoergAtGithub replied:
I personally can't, because I don't know, how it works with command line. Maybe @Holzhaus can describe the command line steps needed for CI.
But anyway, this is out of scope of this PR and should be done in an own PR, or in the Wiki. Scope of this PR is to provide the VisualStudio IDE setup.

Holzhaus replied:
I don't have windows so I can't test this. In theory you should be able to open a VS developer console, `CD` to the mixxx repository, then run:

    C:\mixxx\> tools\windows_buildenv.bat
    C:\mixxx\> MD build
    C:\mixxx\> CD build
    C:\mixxx\> cmake -G Ninja ..
    C:\mixxx\> cmake --build .

daschuer commented on tools/windows_buildenv.bat:
```diff
@@ -95,5 +109,70 @@ EXIT /B 0
+
+:Configuration2CMakeSettings_JSON <optimize> <configurationtype>
+    >>%CMakeSettings% echo     {
+    >>%CMakeSettings% echo       "buildRoot": "${projectDir}\\build_!PLATFORM!__%1",
```
Since we have now removed the build folder from our source tree we can move all buildRoots to build. 
This matches more to the standard cmake approach. 
build\msvc_x64-RelWithDebInfo
or such.

JoergAtGithub replied:
To put all buildRoots to a subdirectory "build" makes sense for me!
The same applies for the installRoots, should the Name of this subdirectory "dist" or "install"?
RelWithDebInfo in the buildRoot would hardcode the configurationType into the directory name, which would be wrong after setting the configurationType-Switch: 
![grafik](https://user-images.githubusercontent.com/64457745/102119902-1968e300-3e42-11eb-8590-d541bcf2cbbd.png)

daschuer replied:
What is the installRoot in sense of cpack?

I think the wix packed is generated by cpack -G WIX 

This will install Mixxx to 

```build/msc_x64-RelWithDebInfi/_CPackBuild/Windows/WIX```

Is this the folder?

JoergAtGithub replied:
> Is this the folder?
I don't have such a folder. installRoot points to a folder where the build generates a clean copy with all files needed for install. Currently I've set it like this:
buildRoot: mixxx/build_x64__fastbuild
installRoot: mixxx/dist_x64__native
AFAIK it's not common to have installRoot as subdirectory of buildRoot. It's usually a parallel path.

daschuer replied:
Ok, than let's stick with the standard.

How does cpack integrate with visual studio? Is it possible to create the installer from this visual studio Gui? Or is it an extra command.

Visual studio calls the configure path from cmake. This generates also the cpack confg files.

Can you verify if this works with the json file? Does cpack uses the installRoot files of does it require a full rebuild of all cpp sources?

JoergAtGithub replied:
I looked for this: CMake and CPack are shipped with VisualStudio and work as intended. But WiX is a third-party packagewhich is not contained in the buildenv-Download. Therefore CPack always fails with:
`CMake Error at C:/Program Files (x86)/Microsoft Visual Studio/2019/Community/Common7/IDE/CommonExtensions/Microsoft/CMake/CMake/share/cmake-3.18/Modules/Internal/CPack/CPackWIX.cmake:12 (message):
  Could not find the WiX candle executable.`

daschuer replied:
Does this work out of the Box using NSIS? Would you mind to try both?

If you use the cmake Visual Studio Generator you get a PACKAGE target. Like described here: 
https://dominoc925.blogspot.com/2016/09/create-windows-installer-using-cmake.html
Do we have similar with the json file?

JoergAtGithub replied:
I investigated this:
1.) WiX needs to be downloaded manually, because it's missing in the Mixxx buildenv download.
2.) Visual Studio ships cpack.exe, but there is no IDE integration for CPack yet
3.) cpack -G WIX works, if called from inside the Visual Studio "Developer Command Prompt"

daschuer replied:
Cool thank you.
Does cpack do a full rebuild?

JoergAtGithub replied:
No, it just creates the .msi file in mixxx\build_x64__fastbuild\_CPack_Packages\win64\WIX

daschuer replied:
Good. So everything should be fine.

JoergAtGithub replied:
I changed buildroot and installroot as sugested
I added the manual steps, to create an .msi installer using WiX

JoergAtGithub committed:
Changed Debug symbols setting
Changed build and install directory names

JoergAtGithub committed:
Changed root paths to ./build and ./install
Changed debug symbol settings
Added WiX installer steps to documentation

daschuer reviewed (approved):
LGTM, thank you.
I cannot test it myself, but since can't break anything, we can merge this.
